src/Microsoft.TemplateEngine.Utils/Glob.cs

```diff
@@ -0,0 +1,275 @@
+                //  we don't need to actually test for a match at this stage
+                //Otherwise, test whether the matcher can consume starting at the current position
+                if (currentMatcher < _matchers.Count - 1 && matcher.ProducesCheckpoint && matcher.MinConsume == 0 || matcher.CanConsume(test, i, out i))
+                {
+                    //If the current matcher isn't that last one and it produces a checkpoint, stash
+                    //  the checkpoint info
+                    if (currentMatcher < _matchers.Count - 1 && matcher.ProducesCheckpoint)
+                    {
+                        checkpoints.Push(new Checkpoint(matcher, currentMatcher + 1, i));
+                    }
+
+                    ++currentMatcher;
+
+                    if (currentMatcher < _matchers.Count)
+                    {
+                        continue;
+                    }
+
+                    if (matcher.ProducesCheckpoint)
+                    {
+                        if (i < test.Length)
+                        {
+                            --currentMatcher;
+                        }
+
+                        continue;
+                    }
+                }
+
+                if (currentMatcher == _matchers.Count && i == test.Length)
+                {
+                    return true;
+                }
+
+                //If the match failed or we ran out of matchers, try to revert to an earlier checkpoint to re-evaluate
+                //If we've got checkpoints left, back up one and see if it's usable
+                while (checkpoints.Count > 0)
+                {
+                    Checkpoint checkpoint = checkpoints.Pop();
+
+                    //If the matcher is usable...
+                    //  restore the checkpoint
+                    //  advance the string position
+                    //  reset the current matcher to the one that followed the checkpoint
+                    if (checkpoint.Matcher.CanConsume(test, checkpoint.StringPosition, out i))
+                    {
+                        checkpoint.StringPosition = i;
+                        currentMatcher = checkpoint.NextMatcher;
+                        checkpoints.Push(checkpoint);
+                        break;
+                    }
+                }
+
+                //If we ran out of checkpoints, the match has failed
+                if (checkpoints.Count == 0)
+                {
+                    return false;
+                }
+            }
+
+            return i == test.Length && currentMatcher == _matchers.Count;
+        }
+
+        private class Checkpoint
+        {
+            public Checkpoint(IMatcher matcher, int nextMatcher, int stringPosition)
+            {
+                Matcher = matcher;
+                NextMatcher = nextMatcher;
+                StringPosition = stringPosition;
+            }
+
+            public IMatcher Matcher { get; }
+
+            public int NextMatcher { get; }
```

Comment:
nit: NextMacthIndex might make it more intuitive that this isn't a matcher. Due to the similarity of the names "Matcher" & "NextMatcher", I assumed they stored the same type of thing until I looked at the implementation.